File: backend/app/models/project.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime, date
from decimal import Decimal
from .common import BaseDocument, Priority, Status, HealthStatus
from enum import Enum

class ProjectType(str, Enum):
    SOFTWARE_DEVELOPMENT = "software_development"
    INFRASTRUCTURE = "infrastructure"
    RESEARCH = "research"
    MARKETING = "marketing"
    PROCESS_IMPROVEMENT = "process_improvement"
    COMPLIANCE = "compliance"
    OTHER = "other"

class ProjectMethodology(str, Enum):
    WATERFALL = "waterfall"
    AGILE = "agile"
    SCRUM = "scrum"
    KANBAN = "kanban"
    HYBRID = "hybrid"
    LEAN = "lean"

class MilestoneStatus(str, Enum):
    PLANNED = "planned"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    DELAYED = "delayed"
    CANCELLED = "cancelled"

class Milestone(BaseModel):
    """Project milestone model"""
    id: str = Field(default_factory=lambda: str(__import__('uuid').uuid4()))
    name: str
    description: Optional[str] = None
    planned_date: date
    actual_date: Optional[date] = None
    status: MilestoneStatus = MilestoneStatus.PLANNED
    deliverables: List[str] = Field(default_factory=list)
    dependencies: List[str] = Field(default_factory=list)  # Other milestone IDs
    
class ProjectFinancials(BaseModel):
    """Project financial information"""
    total_budget: Decimal = Field(default=Decimal('0'))
    allocated_budget: Decimal = Field(default=Decimal('0'))
    spent_amount: Decimal = Field(default=Decimal('0'))
    committed_amount: Decimal = Field(default=Decimal('0'))
    forecasted_cost: Decimal = Field(default=Decimal('0'))
    budget_variance: Decimal = Field(default=Decimal('0'))
    cost_to_complete: Decimal = Field(default=Decimal('0'))
    
    # Cost categories
    labor_cost: Decimal = Field(default=Decimal('0'))
    material_cost: Decimal = Field(default=Decimal('0'))
    vendor_cost: Decimal = Field(default=Decimal('0'))
    overhead_cost: Decimal = Field(default=Decimal('0'))
    
class ResourceAllocation(BaseModel):
    """Resource allocation model"""
    user_id: str
    role: str
    allocation_percentage: float = Field(..., ge=0, le=100)
    hourly_rate: Optional[Decimal] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    skills_required: List[str] = Field(default_factory=list)

class Project(BaseDocument):
    """Project model"""
    name: str = Field(..., max_length=200)
    code: str = Field(..., max_length=50)  # Unique within tenant
    description: Optional[str] = None
    
    # Classification
    project_type: ProjectType
    methodology: ProjectMethodology = ProjectMethodology.AGILE
    status: Status = Status.DRAFT
    health_status: HealthStatus = HealthStatus.GREEN
    priority: Priority = Priority.MEDIUM
    
    # Hierarchy
    portfolio_id: Optional[str] = None
    parent_project_id: Optional[str] = None
    
    # Ownership
    project_manager_id: str
    sponsor_id: Optional[str] = None
    team_members: List[str] = Field(default_factory=list)  # User IDs
    
    # Timeline
    planned_start_date: Optional[date] = None
    planned_end_date: Optional[date] = None
    actual_start_date: Optional[date] = None
    actual_end_date: Optional[date] = None
    
    # Progress tracking
    percent_complete: float = Field(default=0.0, ge=0, le=100)
    milestones: List[Milestone] = Field(default_factory=list)
    
    # Financial information
    financials: ProjectFinancials = Field(default_factory=ProjectFinancials)
    
    # Resource management
    resource_allocations: List[ResourceAllocation] = Field(default_factory=list)
    
    # Risk and issues
    risk_score: float = Field(default=0.0, ge=0, le=1)
    open_issues_count: int = Field(default=0)
    open_risks_count: int = Field(default=0)
    
    # Documents and attachments
    document_urls: List[str] = Field(default_factory=list)
    
    # Custom fields for extensibility
    custom_fields: Dict[str, Any] = Field(default_factory=dict)
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat(),
            date: lambda v: v.isoformat(),
            Decimal: lambda v: float(v)
        }

class ProjectCreate(BaseModel):
    """Schema for creating a new project"""
    name: str = Field(..., max_length=200)
    code: str = Field(..., max_length=50)
    description: Optional[str] = None
    project_type: ProjectType
    methodology: ProjectMethodology = ProjectMethodology.AGILE
    priority: Priority = Priority.MEDIUM
    portfolio_id: Optional[str] = None
    parent_project_id: Optional[str] = None
    project_manager_id: str
    sponsor_id: Optional[str] = None
    planned_start_date: Optional[date] = None
    planned_end_date: Optional[date] = None

class ProjectUpdate(BaseModel):
    """Schema for updating project"""
    name: Optional[str] = None
    description: Optional[str] = None
    project_type: Optional[ProjectType] = None
    methodology: Optional[ProjectMethodology] = None
    status: Optional[Status] = None
    health_status: Optional[HealthStatus] = None
    priority: Optional[Priority] = None
    portfolio_id: Optional[str] = None
    project_manager_id: Optional[str] = None
    sponsor_id: Optional[str] = None
    planned_start_date: Optional[date] = None
    planned_end_date: Optional[date] = None
    actual_start_date: Optional[date] = None
    actual_end_date: Optional[date] = None
    percent_complete: Optional[float] = None
    team_members: Optional[List[str]] = None

class ProjectResponse(BaseModel):
    """Project response model"""
    id: str
    name: str
    code: str
    description: Optional[str]
    project_type: ProjectType
    methodology: ProjectMethodology
    status: Status
    health_status: HealthStatus
    priority: Priority
    portfolio_id: Optional[str]
    project_manager_id: str
    sponsor_id: Optional[str]
    planned_start_date: Optional[date]
    planned_end_date: Optional[date]
    actual_start_date: Optional[date]
    actual_end_date: Optional[date]
    percent_complete: float
    financials: ProjectFinancials
    risk_score: float
    open_issues_count: int
    open_risks_count: int
    team_size: int = 0
    created_at: datetime
    updated_at: datetime

class ProjectDashboard(BaseModel):
    """Project dashboard data"""
    project: ProjectResponse
    milestone_progress: Dict[str, int]
    budget_status: Dict[str, float]
    team_utilization: Dict[str, float]
    recent_activities: List[Dict[str, Any]]
    upcoming_milestones: List[Milestone]

File: frontend/src/types/project.ts
export interface Project {
  id: string;
  name: string;
  code: string;
  description?: string;
  project_type: 'software_development' | 'infrastructure' | 'research' | 'marketing' | 'process_improvement' | 'compliance' | 'other';
  methodology: 'waterfall' | 'agile' | 'scrum' | 'kanban' | 'hybrid' | 'lean';
  status: 'draft' | 'active' | 'on_hold' | 'completed' | 'cancelled';
  health_status: 'green' | 'yellow' | 'red';
  priority: 'low' | 'medium' | 'high' | 'critical';
  current_phase?: 'initiation' | 'planning' | 'execution' | 'monitoring' | 'closure';
  portfolio_id?: string;
  parent_project_id?: string;
  template_id?: string;
  project_manager_id: string;
  sponsor_id?: string;
  team_members: string[];
  planned_start_date?: string;
  planned_end_date?: string;
  actual_start_date?: string;
  actual_end_date?: string;
  percent_complete: number;
  financials: ProjectFinancials;
  risk_score: number;
  team_size: number;
  created_at: string;
  updated_at: string;
}

export interface ProjectFinancials {
  total_budget: number;
  allocated_budget: number;
  spent_amount: number;
  committed_amount: number;
  forecasted_cost: number;
  budget_variance: number;
  cost_to_complete: number;
  labor_cost: number;
  material_cost: number;
  vendor_cost: number;
  overhead_cost: number;
}

export interface ProjectTask {
  id: string;
  name: string;
  description?: string;
  status: 'not_started' | 'in_progress' | 'on_hold' | 'completed' | 'cancelled';
  priority: 'low' | 'medium' | 'high' | 'critical';
  assignee_id?: string;
  reviewer_id?: string;
  planned_start_date?: string;
  planned_end_date?: string;
  actual_start_date?: string;
  actual_end_date?: string;
  estimated_hours?: number;
  actual_hours?: number;
  percent_complete: number;
  parent_task_id?: string;
  subtasks: string[];
  dependencies: ProjectDependency[];
  labels: string[];
  tags: string[];
  story_points?: number;
}

export interface ProjectDependency {
  id: string;
  predecessor_project_id?: string;
  predecessor_task_id?: string;
  successor_project_id?: string;
  successor_task_id?: string;
  dependency_type: 'finish_to_start' | 'start_to_start' | 'finish_to_finish' | 'start_to_finish';
  lag_days: number;
  description?: string;
  is_critical_path: boolean;
}

export interface ProjectIssue {
  id: string;
  title: string;
  description?: string;
  issue_type: 'bug' | 'task' | 'story' | 'epic' | 'improvement' | 'blocker';
  status: 'draft' | 'active' | 'on_hold' | 'completed' | 'cancelled';
  priority: 'low' | 'medium' | 'high' | 'critical';
  reporter_id: string;
  assignee_id?: string;
  created_date: string;
  due_date?: string;
  resolved_date?: string;
  related_task_id?: string;
  blocked_tasks: string[];
  resolution?: string;
  resolution_notes?: string;
  labels: string[];
  estimated_hours?: number;
}

export interface ProjectRisk {
  id: string;
  title: string;
  description: string;
  risk_level: 'low' | 'medium' | 'high' | 'critical';
  probability: number;
  impact: number;
  owner_id: string;
  identified_by: string;
  identified_date: string;
  status: 'draft' | 'active' | 'on_hold' | 'completed' | 'cancelled';
  mitigation_plan?: string;
  contingency_plan?: string;
  mitigation_cost?: number;
  target_resolution_date?: string;
  actual_resolution_date?: string;
  affected_tasks: string[];
  affected_milestones: string[];
}

export interface Milestone {
  id: string;
  name: string;
  description?: string;
  planned_date: string;
  actual_date?: string;
  status: 'planned' | 'in_progress' | 'completed' | 'delayed' | 'cancelled';
  deliverables: string[];
  dependencies: string[];
}

export interface ProjectDetail {
  project: Project;
  tasks: ProjectTask[];
  milestones: Milestone[];
  risks: ProjectRisk[];
  issues: ProjectIssue[];
  dependencies: ProjectDependency[];
  approvals: ProjectApproval[];
  baselines: ProjectBaseline[];
  task_summary: {
    total: number;
    not_started: number;
    in_progress: number;
    completed: number;
    on_hold: number;
  };
}

export interface ProjectApproval {
  id: string;
  approval_type: string;
  status: 'pending' | 'approved' | 'rejected' | 'withdrawn';
  requested_by: string;
  request_date: string;
  description: string;
  justification?: string;
  approver_id: string;
  approval_date?: string;
  approval_comments?: string;
  document_urls: string[];
}

export interface ProjectBaseline {
  id: string;
  name: string;
  description?: string;
  baseline_date: string;
  created_by: string;
  planned_start_date?: string;
  planned_end_date?: string;
  total_budget: number;
  milestones_snapshot: Milestone[];
  tasks_snapshot: ProjectTask[];
  schedule_variance_days?: number;
  cost_variance?: number;
  scope_change_count: number;
}

export interface ProjectCreate {
  name: string;
  code: string;
  description?: string;
  project_type: string;
  methodology?: string;
  priority?: string;
  portfolio_id?: string;
  parent_project_id?: string;
  project_manager_id: string;
  sponsor_id?: string;
  planned_start_date?: string;
  planned_end_date?: string;
}

export interface ProjectIntakeForm {
  project_name: string;
  business_justification: string;
  project_type: string;
  methodology?: string;
  priority?: string;
  requested_start_date?: string;
  target_end_date?: string;
  estimated_budget?: number;
  budget_source?: string;
  project_sponsor: string;
  business_owner: string;
  preferred_project_manager?: string;
  portfolio_id?: string;
  strategic_objectives: string[];
  project_scope: string;
  key_requirements: string[];
  success_criteria: string[];
  estimated_team_size?: number;
  required_skills: string[];
  identified_risks: string[];
  risk_mitigation_notes?: string;
  business_case_url?: string;
  supporting_documents: string[];
}